### ISSUE TITLE
Nullable return type with notnull generic type causes incorrect codegen or incorrect error

### ISSUE THREAD

**Issue by mhutch** [dotnet/roslyn#54807]
**Version Used**:   ce3a20e

**Steps to Reproduce**:

Reduced form below, "real" sample at https://sharplab.io/#gist:bbd5d8798f9b35fb1c28216a8c145337

In summary, things break when a generic method with a `notnull` type parameter returns a nullable version of that type:

```csharp
public static T? NullableDefault<T>()
        where T: notnull
        => default;
```

Returning `default` or `default(T?)` results in `default(T)` rather than the expected `null`.

Trying to explicitly return null results in an incorrect compiler error:

```csharp
public static T? NullableDefault<T>()
        where T: notnull
        => null;
```

```
// Error CS0403  Cannot convert null to type parameter 'T' because it could be a non-nullable value type. 
```

Changing the constraint from `notnull` to either `class` or `struct` causes the method to return `null` as expected. The issue is purely with `notnull`.

**Expected Behavior**:

* The method should always return `null`.
* `default(T?)` should never result in anything other than `null`
* If codegen for this is not possible then the method signature should result in a compiler error that a `notnull` type parameter cannot be used as a nullable return type.

**Actual Behavior**:

When `T` is `int` the method returns `0`.

**Comment by jaredpar** [comment]
> Returning default or default(T?) results in default(T) rather than the expected null.

This behavior is "By Design". The `T?` here is a nullable reference type not a `Nullable<T>`. The `T?` only binds to `Nullable<T>` when either 

1. The `T` is constrained to `struct` in the `where` clause
2. The `T?` exists on a method marked `overrides` where it is not constrained to `default` or `class` (this is purely for back compat reasons). 

That means that `default(T)` is the expected return here because `null` would not be correct from a runtime or language perspective. 

> Trying to explicitly return null results in an incorrect compiler error:

This behavior follows from the idea that the value here is a `T` hence `null` is not a valid value. 

Yes the language could accept `null` here and just translate it to `default(T)` under the hood but doing so would cause other problems. One of the guiding principals of the nullable reference type feature is that adding `?` should not change the semantic meaning of your code. It should only be a mechanism for introducing warnings about accessing potentially `null` values. Allowing a `null` return here would break that principal because it changes the semantic meaning.

A way to visualize how is the following. Essentially if we allow `null` as a return here then we should also by extension allow `null` to be passed as a value for a parameter typed as `T?` when `T: notnull`. That would mean adding a `?` would introduce an overload resolution ambiguity in this code. 

```c#
M(null);
void M(string s) { }
void M<T>(T t) where T : notnull { }
```

When `T?` binds to `Nullable<T>` vs. being a nullable reference type is definitely subtle in a few cases like this. The subtly is something LDM anguished over quite a bit. To the point we originally designed this feature to have the syntax `T??` to make it more obvious. The overwhelming pushback we got from users though is they would rather have `T?` and accept few times where the subtleness of the difference caused friction.

**Comment by mhutch** [comment]
Thanks for the detailed explanation. Is there anything we could do to make this easier to diagnose? I searched for quite a while before filing this and didn't find anything relevant. I understand the problems with making this behave intuitively but it still seems like the compiler could provide better feedback.

Perhaps a warning about `default(T?)` and `default`  when `T` is `notnull`, forcing the user to specify `default(T)` to make behavior clearer?

Perhaps a context-specific variant of `CS0403`? Right now it's warning it can't convert `null` to `T` but I'm trying to return `T?`. Maybe it could tell me why I can't convert `null` to `T?` in this context rather telling me why I can't do something I'm not actually doing?

**Comment by jaredpar** [comment]
> Thanks for the detailed explanation. Is there anything we could do to make this easier to diagnose? I searched for quite a while before filing this and didn't find anything relevant. I understand the problems with making this behave intuitively but it still seems like the compiler could provide better feedback.

@RikkiGibson is this one of the areas you're tracking for better diagnostic support in NRT or should we file an issue to track this?

**Comment by RikkiGibson** [comment]
> Perhaps a warning about `default(T?)` and `default`  when `T` is `notnull`, forcing the user to specify `default(T)` to make behavior clearer?

I would support this change. It feels similar to how we don't allow `expr as string?`--since we don't want you to get the impression that there is a difference from `expr as string`.

I'm also interested in suggestions for how to improve the diagnostic for when the user attempts to convert `null` to `T?` where `T` is not constrained to struct or class. It seems to me like the big shortcoming is that the message tells you "cannot convert null to `T`". Perhaps if we indicated that it's still possible for `T?` to be a non-nullable value type after substitution we could clear it up.

I don't have a specific tracking issue for this so I will open one now.
